src/main/scala/diplomacy/Parameters.scala

```diff
@@ -322,3 +322,44 @@ trait DirectedBuffers[T] {
   def copyOut(x: BufferParams): T
   def copyInOut(x: BufferParams): T
 }
+
+trait IdMapEntry {
+  def name: String
+  def from: IdRange
+  def to: IdRange
+  def isCache: Boolean
+  def requestFifo: Boolean
+  def pretty(fmt: String) =
+    if (from ne to) { // if the subclass uses the same reference for both from and to, assume its format string has an arity of 5
+      fmt.format(to.start, to.end, from.start, from.end, s""""$name"""", if (isCache) " [CACHE]" else "", if (requestFifo) " [FIFO]" else "")
+    } else {
+      fmt.format(from.start, from.end, s""""$name"""", if (isCache) " [CACHE]" else "", if (requestFifo) " [FIFO]" else "")
+    }
+}
+
+abstract class IdMap[T <: IdMapEntry] {
+  protected val fmt: String
+  val mapping: Seq[T]
+  def pretty: String = mapping.map(_.pretty(fmt)).mkString(",\n")
+}
+
+class IdRangeSerial (val start: Int, val end: Int)
```

Comment:
Is this only to be used for the object model? If so, the pattern is to create something like "OMIDRange", similar to `OMAddressSet`: https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala#L18

I'd suggest renaming these to OMIDMap and OMIDRange, respectively, and moving them to
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomaticobjectmodel/model/OMPorts.scala

Without the `_types` the OM deserialization code doesn't know how to deserialize them